src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.c

```diff
@@ -491,7 +495,6 @@ static void pf_connectivity_changed_locked(grpc_exec_ctx *exec_ctx, void *arg,
   }
   if (p->updating_subchannels && error != GRPC_ERROR_NONE) {
     /* Captured the unsubscription for the checking subchannel */
-    GPR_ASSERT(p->selected == NULL);
```

Comment:
Is there any specific reason to remove this assertion? 

Maybe I should ask this first: line 286 implies there may be a selected subchannel when the number of subchannels is 0, I don't know why that can happen?

Reply:
Thanks for detailing this logic.

I have some questions about the race condition when we cancel the notification callback. If we didn't invoke the notification callback with an error to cancel it, how would we cancel it? What will happen if we cancel it in that alternative way when there is already a callback in flight?

Reply:
In C-core, whenever a callback is invoked, there will usually be a window between when the closure is scheduled and when it is actually run.  Note that this can happen in several different ways, depending on which closure scheduler is being used.  For example, with the exec_ctx scheduler, the closure may have been added to the list in the exec_ctx to be executed before gRPC returns control of the thread to the application, but since we are still inside of the gRPC call stack, we have not yet run the closures that were added to the exec_ctx.  Another example is the combiner scheduler, where the closure may have been scheduled on the combiner, but some other thread is currently holding the combiner, so it hasn't been run yet.

Most callbacks hold resources (e.g., refs to some object or allocated memory) that they need to remain alive until after they run.  If we free resources needed by a callback before the callback runs, bad things happen (e.g., crashes or memory corruption).  So the core problem here is that we need to know when we can safely free those resources.

If canceling an async operation did not cause its callback to be invoked, then we would need to free the resources held by the callback immediately upon canceling it, on the assumption that the callback would not ever run.  However, we have no way of knowing at the moment we cancel the operation whether the callback is already pending.  And if it is already pending, then this would result in freeing the resources before the callback runs, which is what we need to avoid here.

Our usual design pattern for solving this problem is to have async operations provide a guarantee that their callback will always be invoked exactly once, regardless of success or failure, so that the callback can free the resources either way.  When we cancel an async operation, there are two cases:
- If the async operation's callback has not yet been scheduled, we schedule it immediately with an error to indicate the cancellation.
- If the async operation's callback has already been scheduled (with no error), the cancellation is a no-op, because the result has already been sent.

Note that in the second case, it's the responsibility of the calling code to know that the operation was cancelled even if the callback is invoked without an error.  Our usual pattern for this is to have the calling code maintain its own `bool shutdown` variable, which it sets to true when cancelling the async operation.  Then, when the callback runs, it knows that there was a problem if either (a) the callback was invoked with an error or (b) the shutdown variable was set to true.

Please let me know if you have any questions about any of this.

Reply:
Mostly understood. I still need to digest some parts and will ask later. Thanks!